### DISCUSSION TITLE
Annotate specific orientation in pole figure using EBSD ID

### DISCUSSION THREAD

**Discussion by GitHubUserWithoutAName** [mtex-toolbox/mtex#1109] [Ask Anything]
Dear MTEX community,

lets say I plot some EBSD data to a pole figure. Then, I use the mouse to click onto an arbitrary data point in the pole figure. This gives me an informational window including the ID, polar angles and Euler angles,

Let's assume the ID is 123 and I want to annotate this specific orientation in the very same pole figure. I go ahead and do:

`id = 123;`
`ori = ebsd('indexed').orientations(id);`
`hold on`
`annotate(ori, 'label', 'data point with id 123');`

This will annotate four poles (in a (111) pole figure, cubic symmetry) as expected, but NOT the point I just selected previously, but some other arbitrary point in the pole figure.

This is very strange, because `ori` DOES have the exact same Euler angles that I get in the preview when selecting the point in the pole figure.

Any ideas what I may be doing wrong? I am plotting to the correct axis (it is the same axis...) and make sure to use the same ebsd variable (i.e. I plotted `ebsd('indexed')` into the pole figure). If I understand the documentation correctly, `annotate` should be able to do what I want, right?

Thanks in advance for any help!

**Comment by GitHubUserWithoutAName** [comment]
I have continued testing this issue unsuccessfully.

At first I thougth this may be related to indixeing via `ebsd('indexed').orientations(id)` versus `ebsd.orientations(id)`, which yields different values (as may be expected). 

I also noticed that after restricting to indexed data, indexing directly is (as is stated in the documentation) not equivalent to actually comparing the id property like: `ebsd(ebsd.id == id)` or `indexedEBSDOnly(indexedEBSDOnly.id == id)` after assigning `indexedEBSDOnly = ebsd('indexed')`. All this is not the problem, however, as both now yield the same result.

Using `indexedEBSDOnly(id)` or `ebsd('indexed').orientations(id)` is the only way to get the EBSD data point with the same Euler angles as the data cursor in the PF returns. This point does however usally possess a different index than is displayed by the data cursor. This means that what is called ID in the data cursor preview is actually not the id property, but just the location of the data point in the ebsd data array, I believe.

Nevertheless, this does not solve the problem: Plotting this point (directly or via annotate), or even defining the orientation from the data cursor manually for a given point always results in a wrong result (i.e. it plots to a completely different location in the PF). T

If I am not completely mistaken, this indicates that the data cursor displays values for a different point than is actually selected. Is this a bug, or am I doing something wrong?

I am using MTEX 5.6.1 on Matlab R2019a, by the way.

Thanks for any help you can offer.

**Comment by GitHubUserWithoutAName** [comment]
I just put together a minimal example using some random mtexdata. Can anyone verify that they get the same behavior? See also the attached image.

The strange thing is that pt has the same Euler angles as the data cursor, but plots to a completely different location.

```
% load some random test data
mtexdata twins
cs = ebsd('indexed').CS;
plotPDF(ebsd('indexed').orientations, Miller(1, 0, 0, cs), 'all');

% I clicked on a random pole and stored the data cursor ID here (see image)
id = 3179;
% display Euler angles of this point: this yields the same values as the data cursor: 4.4°, 80.8°, 195.5°
pt = ebsd('indexed').orientations(id)

% annotate this point: this marks a different point!
annotate(pt, 'label', 'test point', 'MarkerFaceColor', 'red');

% plotting this directly on top yields exactly the same points as annotate - but not the point clicked!
hold on
plotPDF(pt, Miller(1, 0, 0, cs), 'MarkerSize', 5, 'MarkerFaceColor', 'green');
```
![example](https://user-images.githubusercontent.com/77495851/123672895-25e1e580-d840-11eb-9304-34b47ec69e32.PNG)

**Reply by kilir** [reply]
Hi,
I can confirm your observation and this is a bug.
Simple test, plot and click some points:
```
cs = crystalSymmetry('622')
o = equispacedSO3Grid(cs,'resolution',10*degree);
o=discreteSample(o(:),3)
h = Miller(0, 0, 0, 1, cs)
plotPDF(o, h, 'all');
```


Place to start: `tooltip()` https://github.com/mtex-toolbox/mtex/blob/5bb4c1c0bdc77d476bd9e47ccc9fc1583e923b80/geometry/%40orientation/plotPDF.m#L157 which calls `getDataCursorPos()` https://github.com/mtex-toolbox/mtex/blob/develop/plotting/%40mtexFigure/getDataCursorPos.m
I'll look into it as soon as I get some time.


Cheers,
Rüdiger

**Reply by GitHubUserWithoutAName** [reply]
Dear Rüdiger,

thanks a lot for confirming this! I had a look at `tooltip()` and `getDataCursorPos()`. As far as I can tell, `getDataCursorPos()` returns the wrong ID. The return value `pos` seems fine to me (which I have also used successfully in the past when directly calling `getDataCursorPos()`).

In a test case of mine with only 4 orientations plotted, I could trace this back to line 74: ` id = ceil(id*maxId/numel(xd));` in `getDataCursorPos()`. In this test case `maxId = 4`, but the id produced by this line is always 1. This leads to erroneous indexing into `ori` in line 163 in `plotPDF/tooltip()`: `txt{3} = ['Euler = ' char(ori.subSet(id))];` This explains why ID and Euler angles are always consistent but do not actually belong to the point clicked. Ori contains the four correct orienations, as expected.

I do not really understand enough of the inner workings of MTEX to figure out how to obtain the corred ID, unfortunately. `xd` and `yd` in `getDataCursorPos()` both contain a ton of entries in my test case, a lot of which are NaN. 

Line 55:  `[~,id] = min((xd(:)-pos(1)).^2 + (yd(:)-pos(2)).^2);` in `getDataCursorPos()` does however select the "correct" entry as far as I can tell, where `xd` and `yd` are equal to `pos(1)` and `pos(2)` before casting `pos `to `vector3d` (sphericalPlot). The corresponding ID is however much larger than 4, which I gues what `if nargin > 1
  id = ceil(id*maxId/numel(xd));
end` in line 73ff. is for. This however fails to produce the correct ID and always returns 1.

Do you know why `xdata `and `ydata` contain so many NaN values and many more values than poles / orientations in the plot? If I understood what `xdata `and `ydata` entail and how they are constructed, I might be able to figure this out...

**Comment by yuchung-chuang** [comment]
Hi both,

I had a look into this, and I've figured out what's happening.

The issue comes from 
https://github.com/mtex-toolbox/mtex/blob/5bb4c1c0bdc77d476bd9e47ccc9fc1583e923b80/geometry/%40orientation/plotPDF.m#L124
where the orientations `ori` are multiplied by the symmetrised miller indices `sh`, therefore the total number of orientations plotted on the pole figure is equal to `length(ori) * length(sh)`. The order of the orientations is determined by the `reshape()` method.

However, inside `getDataCursorPos()`, the calculation of `id` at
https://github.com/mtex-toolbox/mtex/blob/b6b57b334a8f409d2a76ecc730911c817feb1f7c/plotting/%40mtexFigure/getDataCursorPos.m#L74
does not respect the arrangement by the `reshape()` method. Consequently, we get a wrong `id`.

To fix this, we can change the line to
```matlab
id = mod(id,maxId);
if id == 0
    id = maxId;
end
```
This would give you the right `id` associated to the `ori`. 

The code above has been tested by the following procedure:
```matlab
CS = crystalSymmetry('6/mmm',[2.95 2.95 4.68], [90 90 120]*degree);
m = Miller({0 0 0 1},{1 1 -2 0},CS);
% plot 10 random orientations
ori = orientation.rand(10,CS);
plotPDF(ori, m)
% now click on the figure and get the id showing from the tooltip
% then plot that orientation on the same figure to see if they match
hold on
plotPDF(ori(id), m) 
```

Cheers,
Alex

**Reply by yuchung-chuang** [reply]
Note that if you do something like
```matlab
plotPDF(...)
hold on
plotPDF(...)
```
This will mess up the `id` you get from the cursor, since you're still drawing on the same `mtexFigure`, the `ori` stored in `@tooltip` (which is a closure) doesn't get updated. It doesn't seem to be easy to fix this, as we lose the previous `ori` once we make the second `plotPDF()` call. 
Here's the line worth looking into in case anyone is interested:
https://github.com/mtex-toolbox/mtex/blob/5bb4c1c0bdc77d476bd9e47ccc9fc1583e923b80/geometry/%40orientation/plotPDF.m#L39
